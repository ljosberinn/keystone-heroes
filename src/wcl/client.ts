import type { WCLAuth } from "@prisma/client";
import { GraphQLClient } from "graphql-request";
import fetch from "node-fetch";

import {
  WCL_GQL_ENDPOINT,
  WCL_CLIENT_ID,
  WCL_CLIENT_SECRET,
  WCL_OAUTH_ENDPOINT,
} from "../web/env";
import type { WCLOAuthResponse } from "./auth";
import { getWCLAuthentication, setWCLAuthentication } from "./auth";
import { getSdk } from "./types";
import type { Sdk } from "./types";

type Cache = {
  sdk: Sdk | null;
  client: GraphQLClient | null;
  expiresAt: number | null;
  pending: boolean;
};

const cache: Cache = {
  sdk: null,
  client: null,
  expiresAt: null,
  pending: false,
};

export const getCachedSdk = async (): Promise<Sdk> => {
  if (cache.sdk) {
    return cache.sdk;
  }

  const client = await getGqlClient();
  cache.sdk = cache.sdk ?? getSdk(client);

  return cache.sdk;
};

const FIVE_DAYS_IN_SECONDS = 5 * 24 * 60 * 60;

const mustRefreshToken = (expiresAt: NonNullable<WCLAuth["expiresAt"]>) => {
  const now = Math.floor(Date.now() / 1000);

  if (expiresAt <= now || expiresAt - now <= FIVE_DAYS_IN_SECONDS) {
    return true;
  }

  return false;
};

export const getGqlClient = async (): Promise<GraphQLClient> => {
  if (cache.pending) {
    await new Promise((resolve) => {
      setTimeout(resolve, 50);
    });

    return getGqlClient();
  }

  if (cache.client && cache.expiresAt && !mustRefreshToken(cache.expiresAt)) {
    return cache.client;
  }

  cache.pending = true;
  const persisted = await getWCLAuthentication();

  if (
    persisted?.token &&
    persisted?.expiresAt &&
    !mustRefreshToken(persisted.expiresAt) &&
    !cache.client &&
    !cache.expiresAt
  ) {
    cache.client = new GraphQLClient(WCL_GQL_ENDPOINT, {
      headers: {
        authorization: `Bearer ${persisted.token}`,
      },
      fetch: global.fetch,
    });

    cache.expiresAt = persisted.expiresAt * 1000;
    cache.pending = false;

    return cache.client;
  }

  try {
    const body = new URLSearchParams({
      client_id: WCL_CLIENT_ID,
      client_secret: WCL_CLIENT_SECRET,
      grant_type: "client_credentials",
    }).toString();

    const response = await fetch(WCL_OAUTH_ENDPOINT, {
      body,
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
    });

    if (response.ok) {
      const json: WCLOAuthResponse = await response.json();

      await setWCLAuthentication(json);

      cache.client =
        cache.client ??
        new GraphQLClient(WCL_GQL_ENDPOINT, {
          headers: {
            authorization: `Bearer ${json.access_token}`,
          },
          fetch: global.fetch,
        });

      cache.expiresAt = cache.expiresAt ?? Date.now() + json.expires_in;

      return cache.client;
    }

    throw new Error("unable to authenticate with WCL");
  } catch {
    throw new Error("unable to authenticate with WCL");
  }
};
