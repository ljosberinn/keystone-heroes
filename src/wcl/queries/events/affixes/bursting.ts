import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const BURSTING = {
  damage: 243_237,
  debuff: 240_443,
};

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=4&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20243237%20and%20rawDamage%20%3E%200
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5987472, endTime: 7736379, limit: 2000, filterExpression: "type = \"damage\" and target.type = \"player\" and ability.id = 243237 and rawDamage > 0") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "damage" and target.type = "player" and ability.id = ${BURSTING.damage} and rawDamage > 0`,
  `type in ("applydebuff", "applydebuffstack") and ability.id = ${BURSTING.debuff}`,
];

const isBurstingDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: BURSTING.damage,
});

const isBurstingApplyDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: BURSTING.debuff,
});

const isBurstingApplyDebuffStackEvent =
  createIsSpecificEvent<ApplyDebuffStackEvent>({
    type: "applydebuffstack",
    abilityGameID: BURSTING.debuff,
  });

export const getBurstingEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffEvent | ApplyDebuffStackEvent)[] => {
  if (!affixSet.has(Affixes.Bursting)) {
    return [];
  }

  return [
    ...allEvents.filter(isBurstingDamageEvent),
    ...allEvents.filter(isBurstingApplyDebuffEvent),
    ...allEvents.filter(isBurstingApplyDebuffStackEvent),
  ];
};
