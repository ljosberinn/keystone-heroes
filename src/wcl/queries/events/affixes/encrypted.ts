import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const encryptedDebuffIDs = {
  // Urh
  cdrRegen: 368_239,
  // Wo
  msStealth: 368_241,
  // Vy
  hasteDmg: 368_240,
};

export const encryptedDebuffs = [
  {
    name: "Decrypted Urh Cypher",
    id: encryptedDebuffIDs.cdrRegen,
    icon: "inv_progenitor_anima_blue",
    type: "cdr & regen",
  },
  {
    name: "Decrypted Wo Cypher",
    id: encryptedDebuffIDs.msStealth,
    icon: "inv_progenitor_anima_purple",
    type: "ms & stealth",
  },
  {
    name: "Decrypted Vy Cypher",
    id: encryptedDebuffIDs.hasteDmg,
    icon: "inv_progenitor_anima_green",
    type: "haste & dmg",
  },
];

export const encryptedMinibosses = [
  {
    name: "Urh Dismantler",
    id: 184_911,
  },
  {
    name: "Wo Drifter",
    id: 184_910,
  },
  {
    name: "Vy Interceptor",
    id: 184_908,
  },
];

export const encryptedAbilityIDs = {
  // urh relic
  energyBarrage: 368_077,
  // vy
  shoot: 366_406,
  fusionBeam: 366_409,
  vyCypher: 368_495, // on hit proc
  // Wo
  burst: 366_566,
  stealth: 368_162,
  // Urh
  deconstruct: 366_297,
};

export const encryptedAbilities = [
  {
    name: "Energy Barrage",
    id: encryptedAbilityIDs.energyBarrage,
    icon: "spell_progenitor_missile",
  },
  {
    name: "Shoot",
    id: encryptedAbilityIDs.shoot,
    icon: "spell_progenitor_missile",
  },
  {
    name: "Fusion Beam",
    id: encryptedAbilityIDs.fusionBeam,
    icon: "spell_progenitor_beam",
  },
  {
    name: "Burst",
    id: encryptedAbilityIDs.burst,
    icon: "spell_progenitor_areadenial",
  },
  {
    name: "Deconstruct",
    id: encryptedAbilityIDs.deconstruct,
    icon: "ability_warrior_shieldbreak",
  },
  {
    name: "Wo Cloaking Field",
    id: encryptedAbilityIDs.stealth,
    icon: "ability_vanish",
  },
  {
    name: "Decrypted Vy Cypher",
    id: encryptedAbilityIDs.vyCypher,
    icon: "spell_broker_buff",
  },
];

export const encryptedRelics = [
  {
    name: "Wo Relic",
    id: 185_683,
  },
  {
    name: "Vy Relic",
    id: 185_680,
  },
  {
    name: "Urh Relic",
    id: 185_685,
  },
];

export const filterExpression = [
  `type = "applydebuff" and ability.id in (${Object.values(
    encryptedDebuffIDs
  ).join(", ")})`,
  `type = "damage" and ability.id in (${Object.values(encryptedAbilityIDs).join(
    ", "
  )})`,
];

const isEncryptedDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: Object.values(encryptedDebuffIDs),
});

const isEncryptedDamageOrDamgeTakenEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: Object.values(encryptedAbilityIDs),
});

export const getEncryptedEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Encrypted)) {
    return [];
  }

  return allEvents.reduce<(DamageEvent | ApplyDebuffEvent)[]>((acc, event) => {
    if (
      isEncryptedDamageOrDamgeTakenEvent(event) ||
      isEncryptedDebuffEvent(event)
    ) {
      acc.push(event);
    }

    return acc;
  }, []);
};
