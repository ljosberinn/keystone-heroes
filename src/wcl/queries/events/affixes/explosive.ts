import { Affixes } from "@prisma/client";

import type { AllTrackedEventTypes, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const EXPLOSIVE = {
  unit: 120_651,
  ability: 240_446,
} as const;

export const EXPLOSIVE_HEALTH = 224 as const;

/**
 * @see https://www.warcraftlogs.com/reports/YrA4zyZgQJvHbn32#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24target.id%20%3D120651%20AND%20type%20%3D%20%22damage%22%20AND%20overkill%20%3E%200%5E2%24Off%24%23909049%24expression%24target.type%20%3D%20%22player%22%20and%20rawDamage%20%3E%200%20and%20ability.id%20%3D%20240446
 * ```gql
 * {
 *   reportData {
 *     report(code: "YrA4zyZgQJvHbn32") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 87034, endTime: 1527765, limit: 2000, filterExpression: "(target.type = \"player\" and rawDamage > 0 and ability.name = \"Explosion\") OR (target.name = \"Explosives\" AND type = \"damage\" AND overkill > 0)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `target.type = "player" and rawDamage > 0 and ability.id = ${EXPLOSIVE.ability}`,
  `target.id = ${EXPLOSIVE.unit} and overkill > 0`,
];

const isExplosiveDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: EXPLOSIVE.ability,
});

const createIsExplosiveDeathEvent =
  (targetID: number) =>
  (event: AllTrackedEventTypes): event is DamageEvent => {
    return (
      event.type === "damage" &&
      "targetInstance" in event &&
      event.sourceID !== -1 &&
      event.targetID === targetID &&
      typeof event.overkill !== "undefined" &&
      event.overkill > 0
    );
  };

/**
 * explosives detection is solely based on the hope that it's the unit that
 * occured the most over the course of a logging session, and thus has the highest
 * `targetInstance` property
 */
export const findExplosiveTargetID = (
  allEvents: AllTrackedEventTypes[]
): number | null => {
  const dataset = allEvents.reduce<{
    targetID: null | number;
    targetInstance: null | number;
  }>(
    (acc, event) => {
      if (
        // skip non-DamageEvents
        event.type !== "damage" ||
        // ignore event if target is single instance
        !("targetInstance" in event) ||
        // skip empty instances, just TS here
        event.targetInstance === undefined ||
        // check whether the stored instance is lower than this events
        (acc.targetInstance && event.targetInstance < acc.targetInstance)
      ) {
        return acc;
      }

      return {
        targetID: event.targetID,
        targetInstance: event.targetInstance,
      };
    },
    {
      targetID: null,
      targetInstance: null,
    }
  );

  return dataset.targetID;
};

export const getExplosiveEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): { explosiveTargetID: null | number; events: DamageEvent[] } => {
  if (!affixSet.has(Affixes.Explosive)) {
    return { explosiveTargetID: null, events: [] };
  }

  const explosiveTargetID = findExplosiveTargetID(allEvents);

  if (!explosiveTargetID) {
    return { explosiveTargetID: null, events: [] };
  }

  const explosiveKills = allEvents.filter(
    createIsExplosiveDeathEvent(explosiveTargetID)
  );
  const explosiveDamage = allEvents.filter(isExplosiveDamageEvent);

  return {
    explosiveTargetID,
    events: [...explosiveDamage, ...explosiveKills],
  };
};
