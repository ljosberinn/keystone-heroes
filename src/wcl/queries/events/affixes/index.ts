import { Affixes } from "@prisma/client";

import type { EventParams } from "..";
import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
  RemoveDebuffEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
  | "Infernal"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
  [Affixes.Encrypted]: [],
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): {
  explosiveTargetID: number | null;
  events: (
    | ApplyBuffEvent
    | ApplyBuffStackEvent
    | DamageEvent
    | HealEvent
    | InterruptEvent
    | ApplyDebuffStackEvent
    | BeginCastEvent
    | CastEvent
    | ApplyDebuffEvent
    | RemoveBuffEvent
    | RemoveDebuffEvent
  )[];
} => {
  const affixSet = new Set(affixes);

  const { explosiveTargetID, events: explosiveEvents } = getExplosiveEvents(
    allEvents,
    affixSet
  );

  return {
    explosiveTargetID,
    events: [
      // seasonal
      ...getTormentedEvents(allEvents, affixSet),
      // common affixes
      ...getStormingEvents(allEvents, affixSet),
      ...getSpitefulEvents(allEvents, affixSet),
      ...getSanguineEvents(allEvents, affixSet),
      ...getVolcanicEvents(allEvents, affixSet),
      ...getQuakingEvents(allEvents, affixSet),
      ...getBolsteringEvents(allEvents, affixSet),
      ...getBurstingEvents(allEvents, affixSet),
      ...explosiveEvents,
      ...getGrievousEvents(allEvents, affixSet),
      ...getNecroticEvents(allEvents, affixSet),
    ],
  };
};
