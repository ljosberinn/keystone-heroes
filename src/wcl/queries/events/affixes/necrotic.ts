import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const NECROTIC = 209_858;

/**
 * @see https://www.warcraftlogs.com/reports/DmVgxdj3WTYBApQt/#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24ability.id%20%3D%20209858%20and%20type%20%3D%20%22applydebuffstack%22%20and%20stack%20%3E%2010%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20209858%20and%20rawDamage%20%3E%200
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DmVgxdj3WTYBApQt") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 0, endTime: 1414930, filterExpression: "(target.type = \"player\" and ability.id = 209858) and ((type = \"applydebuffstack\" and stack > 10) or (type = \"damage\" and rawDamage > 0))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */

export const filterExpression = [
  `type in ("applydebuff", "damage", "applydebuffstack") and ability.id = ${NECROTIC}`,
];

const isNecroticDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NECROTIC,
});

const isNecroticDebuffStackEvent = createIsSpecificEvent<ApplyDebuffStackEvent>(
  {
    type: "applydebuffstack",
    abilityGameID: NECROTIC,
  }
);

const isNecroticDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: NECROTIC,
});

export const getNecroticEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (DamageEvent | ApplyDebuffStackEvent | ApplyDebuffEvent)[] => {
  if (!affixSet.has(Affixes.Necrotic)) {
    return [];
  }

  return [
    ...allEvents.filter(isNecroticDebuffStackEvent),
    ...allEvents.filter(isNecroticDamageEvent),
    ...allEvents.filter(isNecroticDebuffEvent),
  ];
};
