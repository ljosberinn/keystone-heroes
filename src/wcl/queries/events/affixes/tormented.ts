import { Affixes } from "@prisma/client";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
} from "../types";

enum TormentedLieutenant {
  SOGGODON_THE_BREAKER = 179_891,
  INCINERATOR_ARKOLATH = 179_446,
  OROS_COLDHEART = 179_892,
  EXECUTIONER_VARRUTH = 179_890,
}

export const tormentedLieutenants = [
  {
    id: TormentedLieutenant.SOGGODON_THE_BREAKER,
    name: "Soggodon the Breaker",
    icon: "inv_icon_wingbroken04a",
  },
  {
    id: TormentedLieutenant.INCINERATOR_ARKOLATH,
    name: "Incinerator Arkolath",
    icon: "ability_warlock_fireandbrimstone",
  },
  {
    id: TormentedLieutenant.OROS_COLDHEART,
    name: "Oros Coldheart",
    icon: "spell_shadow_soulleech_2",
  },
  {
    id: TormentedLieutenant.EXECUTIONER_VARRUTH,
    name: "Executioner Varruth",
    icon: "spell_misc_emotionafraid",
  },
];

export const tormentedLieutenantIDSet = new Set(
  Object.values(TormentedLieutenant).filter(
    (id): id is number => typeof id === "number"
  )
);

export const tormentedSpells = [
  {
    id: 357_524,
    name: "The Stone Ward",
    icon: "ability_mage_shattershield",
    sourceTormentorID: [TormentedLieutenant.SOGGODON_THE_BREAKER],
  },
  {
    id: 356_828,
    name: "Dripping Fang",
    icon: "inv_misc_monsterfang_01",
    sourceTormentorID: [TormentedLieutenant.SOGGODON_THE_BREAKER],
  },
  {
    id: 356_827,
    name: "Tiny Dancing Shoes",
    icon: "inv_boots_pvppriest_e_01",
    sourceTormentorID: [TormentedLieutenant.SOGGODON_THE_BREAKER],
  },
  {
    id: 357_778,
    name: "Broken Mirror",
    icon: "trade_archaeology_highbornesoulmirror",
    sourceTormentorID: [TormentedLieutenant.SOGGODON_THE_BREAKER],
  },
  {
    id: 357_556,
    name: "Self-Embalming Kit",
    icon: "inv_mummypet",
    sourceTormentorID: [TormentedLieutenant.SOGGODON_THE_BREAKER],
  },
  {
    id: 357_864,
    name: "Raging Battle-Axe",
    icon: "ability_hunter_swiftstrike",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_848,
    name: "Signet of Bolstering",
    icon: "ability_socererking_arcanefortification",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_897,
    name: "Crumbling Bulwark",
    icon: "trade_archaeology_stoneshield",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_839,
    name: "The Fifth Skull",
    icon: "inv_misc_bone_skull_02",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_900,
    name: "Bottle of Sanguine Ichor",
    icon: "inv_potion_27",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_889,
    name: "Pedestal of Utter Hubris",
    icon: "ability_paladin_beaconoflight",
    sourceTormentorID: [TormentedLieutenant.INCINERATOR_ARKOLATH],
  },
  {
    id: 357_825,
    name: "Vial of Desperation",
    icon: "trade_archaeology_crackedcrystalvial",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_834,
    name: "Handbook of Uncivil Etiquette",
    icon: "inv_misc_profession_book_cooking",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_817,
    name: "Huntman's Horn",
    icon: "inv_misc_horn_01",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_820,
    name: "Pendant of the Martyr",
    icon: "inv_jewelry_necklace_75",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_842,
    name: "Portable Feeding Trough",
    icon: "inv_crate_07",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_814,
    name: "Regenerative Fungus",
    icon: "inv_misc_starspecklemushroom",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_815,
    name: "Satchel of the Hunt",
    icon: "inv_misc_coinbag11",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_829,
    name: "Gavel of Judgement",
    icon: "inv_hammer_17",
    sourceTormentorID: [TormentedLieutenant.OROS_COLDHEART],
  },
  {
    id: 357_609,
    name: "Dagger of Necrotic Wounding",
    icon: "inv_glaive_1h_maldraxxusquest_b_01",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_706,
    name: "Volcanig Plumage",
    icon: "artifactability_firemage_phoenixbolt",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_863,
    name: "The Stygian King's Barbs",
    icon: "inv_misc_herb_goldthorn_bramble",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_604,
    name: "Siegebreaker's Stand",
    icon: "ability_warrior_unrelentingassault",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_747,
    name: "Overflowing Chalice",
    icon: "inv_drink_22",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_847,
    name: "Stabilizing Diamond Alembic",
    icon: "inv_trinket_80_alchemy01",
    sourceTormentorID: [TormentedLieutenant.EXECUTIONER_VARRUTH],
  },
  {
    id: 357_575,
    name: "Champion's Brand",
    icon: "spell_warrior_sharpenblade",
    sourceTormentorID: [
      TormentedLieutenant.INCINERATOR_ARKOLATH,
      TormentedLieutenant.EXECUTIONER_VARRUTH,
    ],
  },
];

export const TORMENTED = tormentedSpells.map((spell) => spell.id);

/**
 * @see https://www.warcraftlogs.com/reports/J3WKacdjpntmLT7C/#fight=3&type=damage-done&view=events&pins=2%24Off%24%23244F4B%24expression%24(type%20%3D%20%22damage%22%20and%20ability.id%20in%20(357865,%20357841,%20357708,%20357901,%20357525,%20356925,%20356923,%20355806,%20358784,%20358970,%20356667,%20356414,%20358894,%20355709,%20355737,%20358967))%20or%20(type%20%3D%20%22heal%22%20and%20ability.id%20%3D%20357901)
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DrkYtWa2ZPXQ8BjT") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 3895428, endTime: 6047081, filterExpression: "type in (\"applybuff\", \"applybuffstack\") and ability.id in (357524, 356828, 357556, 356827, 357778, 357839, 357575, 357864, 357848, 357889, 357897, 357900, 357814, 357825, 357834, 357815, 357817, 357820, 357842, 357829, 357609, 357706, 357863, 357604, 357747, 357847)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type in ("applybuff", "applybuffstack") and ability.id in (${TORMENTED.join(
    ", "
  )})`,
];
export const tormentedAbilityGameIDSet = new Set(TORMENTED);

export const getTormentedEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (ApplyBuffEvent | ApplyBuffStackEvent)[] => {
  if (!affixSet.has(Affixes.Tormented)) {
    return [];
  }

  return allEvents.reduce<(ApplyBuffEvent | ApplyBuffStackEvent)[]>(
    (acc, event) => {
      if (event.type !== "applybuff" && event.type !== "applybuffstack") {
        return acc;
      }

      if (!tormentedAbilityGameIDSet.has(event.abilityGameID)) {
        return acc;
      }

      // on death, powers get reapplied and we don't care about those
      if (
        acc.some(
          (prevEvent) =>
            event.targetID === prevEvent.targetID &&
            event.abilityGameID === prevEvent.abilityGameID &&
            event.type === prevEvent.type
        )
      ) {
        return acc;
      }

      return [...acc, event];
    },
    []
  );
};
