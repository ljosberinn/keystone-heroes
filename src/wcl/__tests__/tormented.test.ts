import { Affixes } from "@prisma/client";

import { getTormentedEvents } from "../queries/events/affixes/tormented";
import allEvents from "./fixtures/allEvents.json";

describe("getTormentedEvents", () => {
  test("works", () => {
    expect(
      getTormentedEvents(allEvents, new Set([Affixes.Tormented]))
    ).toMatchSnapshot();
  });

  test("does nothing if affix is absent", () => {
    expect(getTormentedEvents(allEvents, new Set())).toHaveLength(0);
  });
});
