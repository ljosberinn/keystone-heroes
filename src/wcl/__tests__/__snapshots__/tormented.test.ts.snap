// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTormentedEvents works 1`] = `
[
  {
    "abilityGameID": 357900,
    "pin": "0",
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 6,
    "targetMarker": 2,
    "timestamp": 4094243,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357897,
    "pin": "0",
    "sourceID": 1,
    "sourceMarker": 7,
    "targetID": 1,
    "targetMarker": 7,
    "timestamp": 4095304,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357575,
    "pin": "0",
    "sourceID": 2,
    "targetID": 2,
    "timestamp": 4096033,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357575,
    "pin": "0",
    "sourceID": 9,
    "targetID": 9,
    "timestamp": 4096613,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357575,
    "pin": "0",
    "sourceID": 8,
    "targetID": 8,
    "timestamp": 4098344,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357817,
    "pin": "0",
    "sourceID": 1,
    "sourceMarker": 7,
    "targetID": 1,
    "targetMarker": 7,
    "timestamp": 4498851,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357814,
    "pin": "0",
    "sourceID": -1,
    "targetID": 9,
    "timestamp": 4500940,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357825,
    "pin": "0",
    "sourceID": -1,
    "targetID": 8,
    "timestamp": 4504750,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357815,
    "pin": "0",
    "sourceID": 6,
    "sourceMarker": 2,
    "targetID": 6,
    "targetMarker": 2,
    "timestamp": 4506129,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357825,
    "pin": "0",
    "sourceID": -1,
    "targetID": 2,
    "timestamp": 4510612,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357706,
    "pin": "0",
    "sourceID": 2,
    "targetID": 2,
    "timestamp": 4741467,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357863,
    "pin": "0",
    "sourceID": 1,
    "sourceMarker": 7,
    "targetID": 1,
    "targetMarker": 7,
    "timestamp": 4741845,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357575,
    "pin": "0",
    "sourceID": 9,
    "stack": 2,
    "targetID": 9,
    "timestamp": 4742709,
    "type": "applybuffstack",
  },
  {
    "abilityGameID": 357575,
    "pin": "0",
    "sourceID": 8,
    "stack": 2,
    "targetID": 8,
    "timestamp": 4743923,
    "type": "applybuffstack",
  },
  {
    "abilityGameID": 357747,
    "pin": "0",
    "sourceID": 6,
    "sourceMarker": 2,
    "targetID": 6,
    "targetMarker": 2,
    "timestamp": 4749497,
    "type": "applybuff",
  },
  {
    "abilityGameID": 356828,
    "pin": "0",
    "sourceID": 1,
    "targetID": 1,
    "timestamp": 5461614,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357524,
    "pin": "0",
    "sourceID": 2,
    "targetID": 2,
    "timestamp": 5463026,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357524,
    "pin": "0",
    "sourceID": 9,
    "targetID": 9,
    "timestamp": 5463652,
    "type": "applybuff",
  },
  {
    "abilityGameID": 357556,
    "pin": "0",
    "sourceID": 6,
    "sourceMarker": 2,
    "targetID": 6,
    "targetMarker": 2,
    "timestamp": 5465915,
    "type": "applybuff",
  },
  {
    "abilityGameID": 356828,
    "pin": "0",
    "sourceID": 8,
    "targetID": 8,
    "timestamp": 5466022,
    "type": "applybuff",
  },
]
`;
